kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/Deduplicated.java

```diff
@@ -0,0 +1,236 @@
+ *       Kafka
+ *       <ul>
+ *         <li>In this case, you should see a log message similar to "Detected 2 data files with the
+ *             same path=data.parquet for table=.* during commit-id=.* in the following
+ *             (deduplicated) events=[Envelope(offset=1)]",
+ *       </ul>
+ *   <li>same file in 2 different envelopes e.g. if a Worker sent the same message twice to Kafka
+ *       <ul>
+ *         <li>In this case, you should see a log message similar to "Detected 2 data files with the
+ *             same path=data.parquet for table=.* during commit-id=.* in the following
+ *             (deduplicated) events=[Envelope(offset=1), Envelope(offset=2)]",
+ *       </ul>
+ *   <li>same file in a single envelope twice e.g. if a Worker included the same file twice in a
+ *       single message in Kafka
+ *       <ul>
+ *         <li>In this case, you should see a log message similar to "Detected 2 data files with the
+ *             same path=data.parquet in the same event=Envelope(offset=1)"
+ *       </ul>
+ * </ul>
+ */
+class Deduplicated {
+  private static final Logger LOG = LoggerFactory.getLogger(Deduplicated.class);
+
+  private Deduplicated() {}
+
+  /**
+   * Returns the deduplicated data files from the batch of envelopes. Does not guarantee anything
+   * about the ordering of the files that are returned.
+   */
+  public static List<DataFile> dataFiles(
+      UUID currentCommitId, TableIdentifier tableIdentifier, List<Envelope> envelopes) {
+    return deduplicatedFiles(
+        currentCommitId,
+        tableIdentifier,
+        envelopes,
+        "data",
+        CommitResponsePayload::dataFiles,
+        dataFile -> dataFile.path().toString());
+  }
+
+  /**
+   * Returns the deduplicated delete files from the batch of envelopes. Does not guarantee anything
+   * about the ordering of the files that are returned.
+   */
+  public static List<DeleteFile> deleteFiles(
+      UUID currentCommitId, TableIdentifier tableIdentifier, List<Envelope> envelopes) {
+    return deduplicatedFiles(
+        currentCommitId,
+        tableIdentifier,
+        envelopes,
+        "delete",
+        CommitResponsePayload::deleteFiles,
+        deleteFile -> deleteFile.path().toString());
+  }
+
+  private static <F> List<F> deduplicatedFiles(
+      UUID currentCommitId,
+      TableIdentifier tableIdentifier,
+      List<Envelope> envelopes,
+      String fileType,
+      Function<CommitResponsePayload, List<F>> extractFilesFromPayload,
+      Function<F, String> extractPathFromFile) {
+    List<FileAndEnvelope<F>> filesAndEnvelopes =
+        envelopes.stream()
+            .flatMap(
+                envelope -> {
+                  CommitResponsePayload payload =
+                      (CommitResponsePayload) envelope.event().payload();
+                  List<F> files = extractFilesFromPayload.apply(payload);
+                  if (files == null) {
```

Comment:
You can get rid of this if statement checking for nulls.  Your tests still pass (the empty ones).  Since you end up doing a `.forEach` on `files` in the next method calls anyways...

Reply:
I don't know how or why it's possible to get nulls there but the null check was there in the original code too so I'm inclined to keep it for now. 

That said; you pointed out a missing test case for me which I've now added [here](https://github.com/tabular-io/iceberg-kafka-connect/pull/212/commits/efc9852f9f8b6037e0c4b0b264de5cbbe4856360).